File: fastapi-python-backend/tests/health_test.py
from fastapi.testclient import TestClient

from src.state.readiness import mark_shutting_down


def test_returns_service_liveness_payload(client: TestClient) -> None:
    response = client.get('/health')

    assert response.status_code == 200
    assert response.json()['success'] is True
    assert response.json()['data']['status'] == 'ok'
    assert response.json()['data']['environment']


def test_returns_ready_state_when_server_is_available(client: TestClient) -> None:
    response = client.get('/ready')

    assert response.status_code == 200
    assert response.json()['success'] is True
    assert response.json()['data']['ready'] is True


def test_returns_503_problem_details_when_shutting_down(client: TestClient) -> None:
    mark_shutting_down()

    response = client.get('/ready')

    assert response.status_code == 503
    assert response.json()['status'] == 503
    assert response.json()['code'] == 'SERVICE_UNAVAILABLE'
    assert '/service-unavailable' in response.json()['type']

File: fastify-ts-bun-backend/tests/health.test.ts
import type { FastifyInstance } from 'fastify';
import { afterEach, describe, expect, it } from 'vitest';
import { createApp } from '../src/app';
import { markShuttingDown, resetReadiness } from '../src/state/readiness';

describe('health and readiness routes', () => {
  let app: FastifyInstance | undefined;

  afterEach(async () => {
    resetReadiness();
    if (app) {
      await app.close();
      app = undefined;
    }
  });

  it('returns service liveness payload', async () => {
    app = await createApp();
    const instance = app;

    const response = await instance.inject({
      method: 'GET',
      url: '/health',
    });

    expect(response.statusCode).toBe(200);
    const body = response.json();
    expect(body.success).toBe(true);
    expect(body.data.status).toBe('ok');
    expect(body.data.environment).toBeDefined();
  });

  it('returns ready state when server is available', async () => {
    app = await createApp();
    const instance = app;

    const response = await instance.inject({
      method: 'GET',
      url: '/ready',
    });

    expect(response.statusCode).toBe(200);
    const body = response.json();
    expect(body.success).toBe(true);
    expect(body.data.ready).toBe(true);
  });

  it('returns 503 problem details when shutting down', async () => {
    app = await createApp();
    const instance = app;
    markShuttingDown();

    const response = await instance.inject({
      method: 'GET',
      url: '/ready',
    });

    expect(response.statusCode).toBe(503);
    const body = response.json();
    expect(body.status).toBe(503);
    expect(body.code).toBe('SERVICE_UNAVAILABLE');
    expect(body.type).toContain('/service-unavailable');
  });
});
